test/test_ddr5.py

```diff
@@ -0,0 +1,144 @@
+        generators = defaultdict(list)
+        generators["sys"].append(dfi.generator(dut.dfi))
+        generators["sys"].append(dfi.reader(dut.dfi))
+        for clock, checker in checkers.items():
+            generators[clock].append(checker.run())
+        pad_generators = pad_generators or {}
+        for clock, gens in pad_generators.items():
+            gens = gens if isinstance(gens, list) else [gens]
+            for gen in gens:
+                generators[clock].append(gen(dut.pads))
+        run_simulation(dut, generators, **kwargs)
+        PadChecker.assert_ok(self, checkers)
+        dfi.assert_ok(self)
+
+    def test_ddr5_empty_command_sequence(self):
+        # Test CS_n/CA values for empty dfi commands sequence
+        latency = '00000000' * self.CMD_LATENCY
+        latency_n = '11111111' * self.CMD_LATENCY
+
+        self.run_test(DDR5SimPHY(sys_clk_freq=self.SYS_CLK_FREQ),
+                      dfi_sequence = [],
+                      pad_checkers = {"sys8x_90": {
+                          'cs_n': latency_n,
+                          'ca0':  latency,
+                          'ca1':  latency,
+                          'ca2':  latency,
+                          'ca3':  latency,
+                          'ca4':  latency,
+                          'ca5':  latency,
+                          'ca6':  latency,
+                          'ca7':  latency,
+                          'ca8':  latency,
+                          'ca9':  latency,
+                          'ca10': latency,
+                          'ca11': latency,
+                          'ca12': latency,
+                          'ca13': latency,
+                      }},
+                      )
+
+    def test_ddr5_ca_addressing(self):
+        # Test that bank/address for different commands are correctly serialized to CA pads
+        latency = '00000000' * self.CMD_LATENCY
+        latency_n = '11111111' * self.CMD_LATENCY
+        read       = dict(cs_n=0, cas_n=0, ras_n=1, we_n=1, bank=0b101, address=0b1100110000)
+        write_ap   = dict(cs_n=0, cas_n=0, ras_n=1, we_n=0, bank=0b111, address=0b10000000000)
+        activate   = dict(cs_n=0, cas_n=1, ras_n=0, we_n=1, bank=0b010, address=0b11110000111100001)
+        refresh_ab = dict(cs_n=0, cas_n=0, ras_n=0, we_n=1, bank=0b100, address=0)
+        precharge_ab  = dict(cs_n=0, cas_n=1, ras_n=0, we_n=0, bank=0b011, address=0)
+        mrw        = dict(cs_n=0, cas_n=0, ras_n=0, we_n=0, bank=0b110011, address=0b10101010)  # bank=6-bit address, address=8-bit op code
+        zqc_start  = dict(cs_n=0, cas_n=1, ras_n=1, we_n=0, bank=0,     address=0b0000101)  # MPC with ZQCAL START operand
+        zqc_latch  = dict(cs_n=0, cas_n=1, ras_n=1, we_n=0, bank=0,     address=0b0000100)  # MPC with ZQCAL LATCH operand
+        mrr        = dict(cs_n=0, cas_n=1, ras_n=1, we_n=0, bank=1,     address=0b101101)  # 6-bit address (bank=1 selects MRR)
```

Comment:
Would be good to have everything indented at the same level, to keep readability

Reply:
Done